### ISSUE TITLE
[RNN] LSTM model crashing during conversion to int8x8 tflite

### ISSUE THREAD

**Issue by CalinPetrea5115** [tensorflow/tensorflow#101050]
### 1. System information

OS: macOS 25.0.0 (ARM64)
Python: 3.10.11
TensorFlow: 2.19.0 from pip
Keras: 3.11.2

### 2. Code
```
import numpy as np
import tensorflow as tf

print(f"TensorFlow: {tf.__version__}")

# Load model (handle legacy LSTM parameter)
class CustomLSTM(tf.keras.layers.LSTM):
    def __init__(self, *args, **kwargs):
        kwargs.pop('time_major', None)
        super().__init__(*args, **kwargs)

with tf.keras.utils.custom_object_scope({'LSTM': CustomLSTM}):
    original_model = tf.keras.models.load_model('conv1dlstm-medium-balanced-0.h5')

input_data = np.load('input_conv1dlstm_medium_balanced_0.npy')

# Create model with fixed batch size
fixed_input = tf.keras.Input(shape=(50, 40), batch_size=1, name='fixed_batch_input')
fixed_output = original_model(fixed_input)
fixed_model = tf.keras.Model(inputs=fixed_input, outputs=fixed_output)

print("1. Float32 TFLite conversion...")
converter_float = tf.lite.TFLiteConverter.from_keras_model(fixed_model)
converter_float.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS]
float_model = converter_float.convert()
print(f"Float32 SUCCESS")

print("2. INT8 quantization...")
converter_int8 = tf.lite.TFLiteConverter.from_keras_model(fixed_model)
converter_int8.optimizations = [tf.lite.Optimize.DEFAULT]
converter_int8.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter_int8.inference_input_type = tf.int8
converter_int8.inference_output_type = tf.int8

def representative_dataset():
    for i in range(50):
        yield [input_data[i:i+1].astype(np.float32)]

converter_int8.representative_dataset = representative_dataset

print("Converting to INT8...")
int8_model = converter_int8.convert()  # SEGMENTATION FAULT HERE
print(f"Int8 SUCCESS")
```

### 3. Failure after conversion
Int8x8 conversion crashes

### 5. (optional) Any other info / logs
If forcing the use of Keras 2 the conversion to both float32 and int8x8 tflite succeeds but the accuracy is trash compared to the original model. With Keras 3 the float32 tflite conversion works like a charm, almost identical accuracy to the original model but int8x8 conversion crashes. The auto-enforcing of training=false when converting with Keras 3 seems to be the difference. Getting the same behavior using Keras 2 and enforcing training=false with a lambda wrapper like this:
```
@tf.function
def call_training_false(x):
    return original_model(x, training=False)
fixed_input = tf.keras.Input(shape=(50, 40), batch_size=1, name='fixed_batch_input')
lambda_output = tf.keras.layers.Lambda(
    lambda x: original_model(x, training=False),
    name='inference_lambda'
)(fixed_input)
lambda_model = tf.keras.Model(inputs=fixed_input, outputs=lambda_output)
```
Expected result => working conversion to int8x8 tflite with Keras 3 or training=false wrapper
Note: This is a classification model, by accuracy I mean matching argmax when comparing tensorflow vs tflite inference outputs

[Archive.zip](https://github.com/user-attachments/files/22545419/Archive.zip)

**Comment by ganesh-karande** [comment]
Hi, I’d like to try investigating and fixing this int8 quantization crash. I’ll start by building a minimal reproducer and digging into the TFLite converter code.

**Comment by CalinPetrea5115** [comment]
@ganesh-karande thank you! let me know if there's any additional info I can provide that helps

**Comment by gaikwadrahul8** [comment]
Hi, @CalinPetrea5115 
Thank you for bringing this issue to our attention, I am able to reproduce the same behavior from my end that [RNN] LSTM model crashing during conversion to int8x8 tflite so we'll have to dig more into this issue and will update you soon. Here is [gist-file](https://colab.sandbox.google.com/gist/gaikwadrahul8/e1619ecaa09f789b4b5e3bbfce52336c/tflite-issue-101050.ipynb) for investigation purpose. Thank you for cooperation and understanding.

```
TensorFlow: 2.19.0
/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
WARNING:absl:Compiled the loaded model, but the compiled metrics have yet to be built. `model.compile_metrics` will be empty until you train or evaluate the model.
WARNING:absl:Error in loading the saved optimizer state. As a result, your model is starting with a freshly initialized optimizer.
1. Float32 TFLite conversion...
Saved artifact at '/tmp/tmpg88lczaa'. The following endpoints are available:

* Endpoint 'serve'
  args_0 (POSITIONAL_ONLY): TensorSpec(shape=(1, 50, 40), dtype=tf.float32, name='fixed_batch_input')
Output Type:
  TensorSpec(shape=(1, 36), dtype=tf.float32, name=None)
Captures:
  134336169363024: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169365904: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169366480: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169365520: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169366288: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169365136: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336175013520: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336175012944: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336175012560: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336175012752: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336175013328: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336175013136: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169368208: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169369936: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169371088: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169368976: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169370512: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169370128: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169370896: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169372240: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169371472: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169373968: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169374352: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169371856: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169371664: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169377040: TensorSpec(shape=(), dtype=tf.resource, name=None)
Float32 SUCCESS
2. INT8 quantization...
Converting to INT8...
Saved artifact at '/tmp/tmp260dyyql'. The following endpoints are available:

* Endpoint 'serve'
  args_0 (POSITIONAL_ONLY): TensorSpec(shape=(1, 50, 40), dtype=tf.float32, name='fixed_batch_input')
Output Type:
  TensorSpec(shape=(1, 36), dtype=tf.float32, name=None)
Captures:
  134336169363024: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169365904: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169366480: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169365520: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169366288: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169365136: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336175013520: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336175012944: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336175012560: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336175012752: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336175013328: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336175013136: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169368208: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169369936: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169371088: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169368976: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169370512: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169370128: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169370896: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169372240: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169371472: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169373968: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169374352: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169371856: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169371664: TensorSpec(shape=(), dtype=tf.resource, name=None)
  134336169377040: TensorSpec(shape=(), dtype=tf.resource, name=None)
```

**Comment by CalinPetrea5115** [comment]
A small update/finding. If I rebuild the model setting unroll=true for the LSTM layer and transfer over the weights the crash is gone and the accuracy seems to be maintained

**Comment by sebastianPsm** [comment]
Hi,

this issue here is related to [#87468](https://github.com/tensorflow/tensorflow/issues/87468) and [#75140](https://github.com/tensorflow/tensorflow/issues/75140).

@tagunil proposed [#76955](https://github.com/tensorflow/tensorflow/pull/76955) (a nullptr check) to solve the issue and it works in general** for me (similar code, [same issue](https://github.com/tensorflow/tensorflow/issues/75140#issuecomment-2341873592)). Unfortunately, the pull request hasn't made it through the review until now and waits in the PR Queue.
** In general, because I needed two more nullptr checks.

How to continue? How about merging the issues 101050, 87468 and 75140?

---
System information:
- OS: Windows + WSL (Ubuntu 24.04) (x86_64)
- TensorFlow: 2.21.0 (CPU, selfbuild)
- Python: 3.12.3

**Comment by CalinPetrea5115** [comment]
Hi,

Getting that merged would be great, but it seems that the PR also failed a check on the AMD ROCm build.

**Comment by aravindmurali22** [comment]
Hi @CalinPetrea5115 , thanks for the PR. 

@suleshahid , could you please take a look at this?

**Comment by aravindmurali22** [comment]
Hi  @CalinPetrea5115 ,
Thank you for taking the time to report this and for your patience.
TFLite has [transitioned](https://developers.googleblog.com/en/tensorflow-lite-is-now-litert/) to LiteRT, Google's primary runtime for on-device AI. All active development, feature updates, and performance enhancements have shifted  to LiteRT. We encourage you to try running your model with LiteRT. If you still encounter this issue, please open a new issue with the updated details in the [LiteRT Repository](https://github.com/google-ai-edge/LiteRT).
For more information and step-by-step guidance on updating your project, please refer to the [LiteRT Migration Guide](https://developers.google.com/edge/litert/migration).
Thank you again!
